### DISCUSSION TITLE
Completely different results (that greatly diverge from each other) by just changing mesh refinement

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#27160] [Q&A General] [answered]
Hi, I am trying to solve the heat equation in the frequency domain. I will try to make this post as detailed as possible, but the tldr is that I've done a lot of debugging and narrowed down the issue to either:

-Incorrect usage of SideSetHeatTransferKernel, or
-Solution method

The input files (and my MOOSE application) referenced in this post can be found here: https://github.com/richmondodufisan/purple/tree/main/FDTR_Simulation_Fourier/3D/Steady_State_Formulation

The heat equation in the diffusive (Fourier) regime is:

![g54](https://github.com/idaholab/moose/assets/52802063/0686e1c9-8722-42d7-aa72-45b57aa79c64)

Taking the Fourier Transform of this, we get:

![g99](https://github.com/idaholab/moose/assets/52802063/d6813b52-8d66-4a06-9d53-ebc654f9f2ae)

To solve this problem numerically, we need to decompose into a coupled system of equations. This is where I discovered the first issue. By decomposing "T hat" into U + iV, we can get a system of two equations. However, when I implemented this, while the results were close to the analytical solution, I saw that there was a "forward" phase shift instead of a phase lag as expected (i.e phase difference between a modulated heat source and the temperature increase). I fixed this by changing the decomposition to U - iV (which I know is questionable, but it worked), such that the system of equations becomes:

![g5](https://github.com/idaholab/moose/assets/52802063/cae3bcd3-c771-4d3d-89be-914f6691fa65)

and

![g54111](https://github.com/idaholab/moose/assets/52802063/3321d95e-7f16-4692-8d39-44c6b62a14f9)

I ran a couple of simulations for the `SingleLayer_Test.i` input file, varying material properties and they all matched with the analytical solution (https://colab.research.google.com/drive/13vhZWXeZJLbZuwk0OFUb_Qxw-fBa7lZ7?usp=sharing) pretty well, at least up to the 3rd decimal place. The exodus output also looked fine, we clearly see diffusive behavior at the location of the heat source (below is the real part of the solution):

![single_real](https://github.com/idaholab/moose/assets/52802063/fb758854-07d6-446a-a8bd-630f3b0c3886)

Next, I tried to extend this to the problem I am actually trying to solve, which consists of two layers with an interface between them. I tried for one mesh refinement, and while the phase calculated was somewhat close, the amplitude was off almost by a factor of 2. The exodus output also looked a bit peculiar, with a weird zone where it should be normal diffusive behavior:
 
![FDTR_real](https://github.com/idaholab/moose/assets/52802063/b6d365c9-7391-4cf6-adb2-af00121d073f)

Since I don't really care about the amplitude for my purposes (even though it should have been a warning sign), I tried to refine the mesh to get closer to the analytical phase (https://colab.research.google.com/drive/1D2EzYUD_GQuQCWPht3R5EaZdGTBTuJn-?usp=sharing). And the result was just completely off. 

![confused](https://github.com/idaholab/moose/assets/52802063/e2947643-e058-42ac-8785-e79ba77af75a)

This left me baffled, because I would expect that if it were an issue of mesh refinement, the results would be somewhat close, still. I used the same further refined mesh for the single layer model and it converged to the analytical solution! 

![singlemodelconverge](https://github.com/idaholab/moose/assets/52802063/51ff0b59-fd09-404b-b34c-02752101bacb)

Which led me to believe that there is either something wrong with my use of SideSetHeatTransferKernel, or the solution method- maybe MOOSE is having trouble finding a unique solution.

I used ADKernel to implement the model, and switched to a Finite Difference Preconditioner. I also changed the pc-type to "lu" as suggested by @lindsayad when I had a problem with convergence issues in the past. Since I have eliminated the kernel as the culprit (since it is clearly giving the correct solution for a single layer model with no interface), I tried switching the "var" and "neighbor_var" variables around for SideSetHeatTransferKernel (don't think they should be impacted, but since the sign of the BC term in the finite element discretization changed after I switched the decomposition to U - iV, maybe the normal is defined differently than I thought or something) to no success.

I am running out of ideas and actually am pretty floored as I've never run a problem where I get completely different results for the same mesh with different refinements, even fracture problems at least show similar patterns. The visual exodus output is a hint that something is not quite right, but I am struggling to think of where else to look. Is there anything I haven't thought of yet that might be worth looking into?

**Comment by GiudGiud** [comment]
@lindsayad I ll let you dispatch this.

**Reply by richmondodufisan** [reply]
@GiudGiud @lindsayad An update to this, I changed the solve type to FD. By my understanding that should be the most accurate way of constructing the Jacobian possible, basically equivalent to an analytical Jacobian if the mesh is fine enough, right? It took 21 hours to solve (as opposed to ~10 minutes with NEWTON, so even if it worked it's unfeasible for practical use), but it still gave me the same result. **Does this rule out the solution method as the cause of the error?** I was only able to solve for one mesh (because of how long it takes to solve).

**Reply by GiudGiud** [reply]
Hello

Yes it s an indication that the Jacobian is correct. I dont think you need to use FD more.
I think something must be wrong with the problem set up. Can you try an intermediate case between the one thsat worked and the one that did not that would isolate one feature?

Guillaume

**Reply by richmondodufisan** [reply]
If the Jacobian is correct, it means the only difference between both problems is the use of SideSetHeatTransferKernel at the interface of the two blocks. Everything else is the same, all boundary conditions and geometry. I guess an intermediate case would be no jump in temperature between each block. I'll give that a shot.

**Reply by richmondodufisan** [reply]
I just ran a two-layer problem with no temperature jump. The results also matched the analytical solution. Which brings me to the conclusion that the issue is with the use of SideSetHeatTransferKernel. Why it isn't working is still a puzzle to me (and why it's producing different results for the same mesh with different refinements)

![two_layer](https://github.com/idaholab/moose/assets/52802063/7594ddda-ec12-4ceb-bbf0-79247ed2763b)

**Comment by lynnmunday** [comment]
In the two layer problem, do you get the same answer with mesh refinement when you apply the load to the top surface (sample_pump_area) instead of the interface?  

This is a different problem but it there might be parts of it you can use:  https://mooseframework.inl.gov/modules/solid_mechanics/1d_elastic_waves.html

**Reply by richmondodufisan** [reply]
Hi, the load actually is applied to the top surface (top_pump_area), which is at the top of the first (very thin, 0.09 microns) layer. "sample_pump_area" is actually the interface, it was only named such because it is at the top of the sample (second layer), and it used to be just a "portion" of the interface and not the whole surface. in a different version of this input file i had a postprocessor to check the temperature increase at the interface as well.

i'll check out the linked page, thanks!

**Comment by richmondodufisan** [comment, marked as answer]
SOLVED!

I mentioned in the original question that for some reason, I had to change the decomposition to U - iV instead of U + iV. I still do not have a mathematically rigorous reason for that, but I guess I'll figure that out later.

When I got the wrong results, I tried switching around the `var` and `neighbor_var` in SideSetHeatTransferKernel for the imaginary and real parts individually (and together), thinking that the change of the sign in the decomposition could be the reason. Apparently this does not have the same effect as what eventually solved it- the solution was to **_change the interface conductance to a negative value_** (and leave the `var` and `neighbor_var` as they were). I had looked at the source code for SideSetHeatTransferKernel and didn't consider doing this, because the way the jump was calculated in computeQpresidual, switching `var` and `neighbor_var` around should have the same effect. So I'm guessing that the error is coming from the Jacobian computed by that Kernel, since it is an analytical Jacobian (not AD). That's the only explanation I could think of. Still doesn't 100% make sense to me, especially since I got extremely different results with the same mesh at different refinements. But it's solved now and I'm relieved! Same result for both meshes!

@GiudGiud @lynnmunday @lindsayad

**Reply by lindsayad** [reply]
A wrong Jacobian should not affect the final result as long as the solver has converged and there is a unique solution
